z3/src/z3_solver.cpp

```diff
@@ -252,33 +353,303 @@ Sort Z3Solver::make_sort(const Sort &sort_con, const SortVec &sorts) const {
+		}
+		res = Z3_mk_rotate_left(ctx, op.idx0, zterm->term);
+	} else if (op.prim_op == Rotate_Right) {
+		if (op.idx0 < 0) {
+			throw IncorrectUsageException("Can't rotate by negative number");
+		}
+		res = Z3_mk_rotate_right(ctx, op.idx0, zterm->term);
+	} else if (op.prim_op == Int_To_BV) {
+		if (op.idx0 < 0) {
+			throw IncorrectUsageException(
+					"Can't have negative width in Int_To_BV op");
+		}
+		res = Z3_mk_int2bv(ctx, op.idx0, zterm->term);
+	} else if (op.prim_op == BV_To_Nat) {
+		// n.b., the third parameter is a boolean is_signed, by flagging it false, this becomes bv2nat
+		res = Z3_mk_bv2int(ctx, zterm->term, false);
+	}
+
+	else if (!op.num_idx) {
+		if (unary_ops.find(op.prim_op) != unary_ops.end()) {
+			res = unary_ops.at(op.prim_op)(ctx, zterm->term);
+		} else {
+			string msg("Can't apply ");
+			msg += op.to_string();
+			msg += " to the term or not supported by Z3 backend yet.";
+			throw IncorrectUsageException(msg);
+		}
+	} else {
+		string msg = op.to_string();
+		msg += " not supported for one term argument";
+		throw IncorrectUsageException(msg);
+	}
+
+	return std::make_shared < Z3Term > (to_expr(ctx, res), ctx);
 }
 
 Term Z3Solver::make_term(Op op, const Term &t0, const Term &t1) const {
-	throw NotImplementedException(
-			"Term iteration not implemented for Z3 backend.");
+	shared_ptr<Z3Term> zterm0 = static_pointer_cast < Z3Term > (t0);
+	shared_ptr<Z3Term> zterm1 = static_pointer_cast < Z3Term > (t1);
+	Z3_ast res;
+
+	if (zterm0->is_function || zterm1->is_function){
+		throw IncorrectUsageException("Cannot make a binary op term with a function.");
+	}
+
+	check_context(zterm0->term, zterm1->term);
+
+	if (!op.num_idx) {
+		if (binary_ops.find(op.prim_op) != binary_ops.end()) {
+			res = binary_ops.at(op.prim_op)(ctx, zterm0->term, zterm1->term);
+		} else if (variadic_ops.find(op.prim_op) != variadic_ops.end()) {
+			Z3_ast terms[2] = { zterm0->term, zterm1->term };
+			res = variadic_ops.at(op.prim_op)(ctx, 2, terms);
+		} else {
+			string msg("Can't apply ");
+			msg += op.to_string();
+			msg += " to the term or not supported by Z3 backend yet.";
+			throw IncorrectUsageException(msg);
+		}
+	} else {
+		string msg = op.to_string();
+		msg += " not supported for two term argument";
+		throw IncorrectUsageException(msg);
+	}
+
+	return std::make_shared < Z3Term > (to_expr(ctx, res), ctx);
 }
 
 Term Z3Solver::make_term(Op op, const Term &t0, const Term &t1,
 		const Term &t2) const {
-	throw NotImplementedException(
-			"Term iteration not implemented for Z3 backend.");
+	shared_ptr<Z3Term> zterm0 = static_pointer_cast < Z3Term > (t0);
+	shared_ptr<Z3Term> zterm1 = static_pointer_cast < Z3Term > (t1);
+	shared_ptr<Z3Term> zterm2 = static_pointer_cast < Z3Term > (t2);
+	Z3_ast res;
+
+	if (!op.num_idx) {
+		if (zterm0->is_function || zterm1->is_function || zterm2->is_function){
+			throw IncorrectUsageException("Cannot make a ternary op term with a function.");
```

Comment:
Just FYI, you actually could do this. For example: `make_term(Apply, f, x, y)`. However, this isn't important to fix now.  I wouldn't be surprised if other solver backends don't handle that case either. Instead they just assume that function applications always use the `make_term(Op, TermVec)` signature.